0000-ldap-sync-tool.md

```diff
@@ -0,0 +1,139 @@
+- Feature Name: `mgr-ldapsync`: LDAP sync tool
+- Start Date: 2019, Oct 28
+- RFC PR: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+The `mgr-ldapsync` is a tool that performs synchronisation of the
+users between Uyuni server and LDAP.
+
+# Motivation
+[motivation]: #motivation
+
+Uyuni server can authenticate users from LDAP via PAM. For this, users
+needs to be added to Uyuni with the flag "PAM authentication" and thus
+their credentials will be verified via LDAP directory. Synchronisation
+tool should transparently synchronise users from certain groups and/or
+roles in LDAP and reflect that in Uyuni server, overriding local
+changes.
+
+Often LDAP configuration at infrastructures is unpredictable. This
+tool should bring certain boundaries and over up most common LDAP
+setups for users and their roles management.
+
+# Detailed design
+[design]: #detailed-design
+
+The tool should work as an aggregator, collecting users from LDAP and
+Uyuni server synchronising them.
+
+## Functionality Aspects
+
+The following functions should be implemented:
+
+- Distinguish between local users in Uyuni and LDAP users.
+- Prevent UID clashes between the local users and LDAP users.
+- Prevent accidental modification or removal of the local user.
+- Flexible support various LDAP schemas, such as eDirectory, POSIX,
+  Active Directory etc.
+- Role management.
+- Dry run mode (preview only).
+
+## Configuration
+
+Configuration should be default in a separate file
+`/etc/rhn/ldapsync.conf`. Alternatively, user should be able to
+specify own path, in case the default one is not created or used.
+
+The main principle of the tool is to map specific roles and or groups
+in LDAP to the roles in the Uyuni server via the configuration. Adding
+or removing LDAP users from these groups or roles should reflect in
+Uyuni without additional manual changes in the Uyuni server.
+
+Configuration should be just a YAML file and the mapping should
+implement a relation between LDAP group to a set of roles from
+Uyuni. So each user in LDAP that is a member of such group would
+inherit these roles. For example:
+
+```yaml
+directory:
+  groups:
+    cn=sysop,ou=groups,dc=example,dc=com:
+      - activation_key_admin
+      - channel_admin
+      - config_admin
+```
```

Comment:
I think this example is incomplete. You say below, that the class is `groupOfNames` and there should be a `member` attribute which specify the user.

Other schemas uses other classes and other attributes. I think the config need to reflect this and the tool needs to work with these different schemas.

Maybe complete examples of the 2 or 3 most used schemas would be a good idea to have better overview what we need to implement. 

- AD
- POSIX
- SUSE?

Reply:
> Other schemas uses other classes and other attributes. 

Yes, AD (which is not POSIX) is using `group` instead. That is simply included in the search filter making literally no difference by default, but I can still mention that if you want.

I don't know anything about SUSE schema (it should be still POSIX), since I never found a documentation of it, but only followed standards of Active Directory and POSIX. If you could provide examples of it, then it would be very helpful.

Reply:
@mcalmer I've added [Attribute Remapping section](https://github.com/uyuni-project/uyuni-rfc/pull/23/commits/bf10f44b5ab179ffb51846d5a28c62ea8c30aa1f) specifically for this. That would allow support custom schemas.

Reply:
@mcalmer  [New commit ID](https://github.com/uyuni-project/uyuni-rfc/pull/23/commits/37490a2cda3f564169b81b236d3ce7bce4e9d0e4) for the [Attribute Remapping](https://github.com/uyuni-project/uyuni-rfc/pull/23/commits/37490a2cda3f564169b81b236d3ce7bce4e9d0e4) section (force-pushed)...